mParticle-Google-Analytics-Firebase/MPKitFirebaseAnalytics.m

```diff
@@ -207,12 +208,61 @@ - (MPKitExecStatus *)logEvent:(MPEvent *)event {
         return [self execStatus:MPKitReturnCodeFail];
     }
     
+    event.name = [self standardizeNameOrKey:event.name forEvent:YES];
+    event.info = [self standardizeValues:event.info forEvent:YES];
     [FIRAnalytics logEventWithName:event.name
                         parameters:event.info];
     
     return [self execStatus:MPKitReturnCodeSuccess];
 }
 
+- (NSString *)standardizeNameOrKey:(NSString *)nameOrKey forEvent:(BOOL)forEvent {
+    NSString *standardizedString =  [nameOrKey stringByReplacingOccurrencesOfString:@"  " withString:@"_"];
+    NSCharacterSet *notAllowedChars = [[NSCharacterSet characterSetWithCharactersInString:@"_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"] invertedSet];
+    standardizedString = [[standardizedString componentsSeparatedByCharactersInSet:notAllowedChars] componentsJoinedByString:@""];
+    if (standardizedString.length > 9 && [standardizedString hasPrefix:@"firebase_"]) {
+        standardizedString = [standardizedString substringFromIndex:9];
+    } else if (standardizedString.length > 7 && [standardizedString hasPrefix:@"google_"]) {
+        standardizedString = [standardizedString substringFromIndex:7];
+    } else if (standardizedString.length > 2 && [standardizedString hasPrefix:@"ga"]) {
+        standardizedString = [standardizedString substringFromIndex:2];
+    }
+    
+    NSCharacterSet *letterSet = [NSCharacterSet characterSetWithCharactersInString:@"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"];
+    
+    while (![letterSet characterIsMember:[standardizedString characterAtIndex:0]]) {
+        standardizedString = [standardizedString substringFromIndex:1];
+    }
+    
+    if (forEvent) {
+        standardizedString = [standardizedString substringToIndex:39];
```

Comment:
What is the meaning of these magic numbers? (39/23)

I know the other ones above are string lengths. I think it would be best to define the strings separately as constants and call length on them rather than having the risk that the number of characters in the literal string could diverge from the lengths, even though that's not the case currently.

Similar for 99/35 below, not sure the meaning but if you make them variables with names it should be easier to understand.

Reply:
@peterjenkins Take a look at the Target Process for this PR. I'd link it but this is public. These are firebase constants/limitations.